doc/developer/design/20231102_read_holds.md

```diff
@@ -0,0 +1,154 @@
+It cannot exist on sinks or unmaterialized views.
+If a `HOLD` exists for a directly queryable object (i.e., not indexes) at time `t`, `SELECT` and `SUBSCRIBE` on that object `AS OF t` will successfully execute.
+If a `HOLD` exists for an index, the same is true if the index satisfies the query.
+A `HOLD` on any non-index object will not also create the `HOLD` on those indexes.
+Because this may be surprising to users, if an `AS OF` query whose plan uses an index errors because the `AS OF` is less than the index's read frontier and a `HOLD` exists on the index's object,
+a helpful error message will be produced telling the user that their `HOLD` must be on the index, not the underlying object.
+
+The objects the `HOLD` is created on do not need to be in the same schema or database.
+The `HOLD` can be created in any schema on which the user has a `CREATE` privilege (see (Alternatives)[#Alternatives] for other takes on this).
+The user must have a `READ` privilege on all objects.
+
+Definitions:
+- **physical read frontier**: the read frontier of an object excluding `HOLD`s and any running queries or transactions.
+- **logical read frontier**: the read frontier of an object including `HOLD`s and any running queries or transactions.
+
+If `AT <time>` is specified, the read hold is created at that time.
+It is an error if that time is less than any object's logical frontier.
+Due to the use of the logical frontier, a new `HOLD` can be created while at the time of an existing `HOLD` (say, because the object set of a `HOLD` needs to be changed by creating a new one).
+If `AT <time>` is not specified, Materialize automatically chooses the least common physical read frontier of all objects.
+
+`HOLD`s are destroyed by:
+
+```
+DROP HOLD <name>
+```
+
+A `DROP` will release its current read hold.
+In-progress `SUBSCRIBE`s and transactions will not be affected after a `DROP`, but will continue to run as expected.
+Those will still be aquiring their own read holds in addition to the `HOLD` object.
+
+Creating a `HOLD` will increase resource consumption (mostly memory) in proportion to the amount of time it is held for.
+Users must periodically advance the `HOLD` to a more recent time to allow for resources to be released.
+
+`HOLD`s are advanced by:
+
+```
+ALTER HOLD <name> ADVANCE [TO <time>]
+```
+
+If `TO <time>` is specified, the new time will be created at that time.
+It is an error if that time is less than any object's logical read frontier.
+If `TO <time>` is not specified, the new time is the least common physical read frontier of all objects.
+The `HOLD` releases its current read hold and aquires a new read hold at the new time.
+
+The `ERROR_IF_OLDER_THAN` option (required, but defaults to `'24h'`) is a duration.
+If difference between the least common physical read frontier of the objects and the `HOLD` time is greater than `ERROR_IF_OLDER_THAN`, the `HOLD`'s read hold is released and the `HOLD` enters an error state.
```

Comment:
"the least common physical read frontier of the objects" seems important here; what does it mean though? Should this be *write* frontier, instead?

Reply:
If so, it seems more ergonomic to indicate "I would like this hold to lag the write frontier by `duration`" rather than ask for repeated manual advancement and errors if you don't make it in time.

Reply:
(Which is what we call a "custom compaction window" I think!)